_posts/2018-07-17-its-a-lot-more-snappy.md

```diff
@@ -0,0 +1,174 @@
+---
+layout: post
+title: "It's A Lot More Snappy"
+date: 2018-07-17
+categories: PWA SPA JavaScript
+author: Rick Foxcroft
+image: img/repeatable-serverless-applications/less_servers.png
+---
+
+**It's A Lot More Snappy**
+_‘A real world application of PWA technology’_
+
+Recently we undertook the journey of building a PWA(progressive web app) for a client here in Perth.  The app, is a maintenance app that allows maintenance technicians to raise issues with equipment in geologically  isolated areas. Due to this isolation, it was important that app could function in an offline mode and to do so, the app needed to store some data on the device it was running on.
+
+Whilst building the app, we encountered some performance issues. Over time the app would gradually become unresponsive to the point where, it would take a few seconds to register a button click.
+
+## What Was The Problem?
+
+* The app had to deal with a somewhat large dataset. One of these datasets was 205509 objects to be exact. Let’s refer to these objects as Pokemon.
+
+* The list of Pokemon had to be used in several places throughout the app and in some places not all of the data pertaining to each Pokemon was used.
+
+* A JSON file of this magnitude takes a fair amount of time to download.
+
+## The Solution:
+
+* _Remove extraneous Data_ from the array and ensure this operation only happens once
+
+* _Create a seperate thread to update indexedDB_, while the user progressed with other aspects of the application.
+
+* _Improve the download speed of the data for slow networks_ by chunking requests in parallel
+_(optional and beyond the scope of this post)_
+
+1. *Remove extraneous Data*
+
+```typescript
+async function transformPokemon(): Promise<void> {
+    for (let i = 0; i < this.Pokemon.length; i++) {
+        const pokemon = {
+            value: this.pokemon[i].pokemonId,
+            label: `${this.pokemon[i].pokemonId} - ${this.pokemon[i].shortDescription}`
+        };
+        this.pokemonOptions.push(pokemon);
+    }
+    await localforage.setItem(this.pokemonGroup, this.pokemonOptions);
+}
+```
+
+Instead of storing the complete dataset in memory and using it in several different places. We pulled out only the data that was absolutely necessary for displaying in the UI. We also made sure that this array mapping operation only happened once every time the UPDATE_POKEMON action was dispatched to our redux store, instead of in a component render method. You'll notice, we also store a copy of this transformed array so that we don't have to retrieve it over the network the next time. Which means we can run the following when mounting our component.
```

Comment:
Mostly good. You've spoken in two different tenses though. You said "We did x", "We also made sure" then you say "so we don't", "which means we can". It's not necessarily wrong, but in its current form reads strange.